### PR TITLE
FOCUS #547: Consistency Review - Standardize column specs for correction charges

### PR DESCRIPTION
**Pull request by ijurica**
Standardize the column specifications, i.e. allow nulls in the case of corrections for all columns except cost metrics, and explicitly define value ranges for all metrics, for the following columns:

- PricingUnit
- ConsumedUnit
- CommitmentDiscountUnit
- PricingQuantity
- ConsumedQuantity
- CommitmentDiscountQuantity
- ListCost
- ContractedCost
- ListUnitPrice
- ContractedUnitPrice

---

### NOTE:

An initial overall analysis of the current correction handling and value ranges was conducted within this PR. 
Due to deadlines, **the scope of this particular PR was limited to addressing the following two columns**:

- **ConsumedUnit**
- **ConsumedQuantity**

Please use this [table](https://docs.google.com/spreadsheets/d/1SvaD-PSzxwRKExKYbQXbtEmVPtbhN4HnycOatvLfGPI/edit?gid=1823426118#gid=1823426118) to verify the normative requirements.

### CODE REVIEW

**Comment by ijurica** [comment]
**PRIORITIES:**

---

- **Correction charges and nullability** – highest priority
- **Explicit definition of value ranges** – lower priority
  - A table overview available in [#547: Corrections and value ranges - Metrics and ValueRanges sheet](https://docs.google.com/spreadsheets/d/1SvaD-PSzxwRKExKYbQXbtEmVPtbhN4HnycOatvLfGPI/edit?gid=1728767078#gid=1728767078) and should be used as the basis for discussion and consensus
- **Switch to bullet list and improve readability** – lowest priority (probably version 1.2)

---

**STATUS:**

---

- **ListUnitPrice and ContractedUnitPrice**
  - Nullability and value ranges correctly specified.
  - Switch to bullet list – on hold.
  - Assumption: AS-IS for 1.1

---

- **Cost columns**
  - **Nullability:** Already specified that these columns MUST NOT be null.
  - **Value ranges** – on hold.
    - Discuss and confirm before proceeding with the PR (table overview available in [#547: Corrections and value ranges - Metrics and ValueRanges sheet](https://docs.google.com/spreadsheets/d/1SvaD-PSzxwRKExKYbQXbtEmVPtbhN4HnycOatvLfGPI/edit?gid=1728767078#gid=1728767078)).
    - Suggestion: Switch to bullet list prior to addressing value ranges.
  - Switch to bullet list – on hold.
  - Assumption: AS-IS for 1.1

---

- **ConsumedQuantity and ConsumedUnit**
  - Bullet list just introduced in this version.
  - **Identified issues/improvement suggestions:**
    - The statement **"ConsumedUnit MUST NOT be null if ChargeCategory is 'Usage' and ChargeClass is not 'Correction'"** is incorrect.
  - Alternative version of both columns, with value ranges explicitly mentioned, are provided in this PR.

---

- **CommitmentDiscountUnit and CommitmentDiscountQuantity**
  - Bullet list already used and value range explicitly mentioned.
  - **Identified issues/improvement suggestions:**
    - Specify that even in the case of corrections...
    - Encompass the fact that corrections are expected/allowed for purchases as well.
    - Specify that positive values are expected for regular charges, while any value is allowed in case of corrections.
  - Two alternative versions are available: one in this PR and another in PR #550.
    - In this PR, the bullets are reordered and a bit rephrased for better readability.
    - PR #550 contains minimal adjustments.
  
---

- **PricingQuantity and PricingUnit**
  - Nullability correctly specified.
  - **Identified issues/improvement suggestions:**
    - The following statement doesn't specify much: "The value MAY be negative in cases where ChargeClass is 'Correction.'"
       By the way, it’s assumed that negative values are expected in regular credit charges (see [#547: Corrections and value ranges - Metrics and ValueRanges sheet](https://docs.google.com/spreadsheets/d/1SvaD-PSzxwRKExKYbQXbtEmVPtbhN4HnycOatvLfGPI/edit?gid=1728767078#gid=1728767078) for more details).
    - Switch back to the version 1.0 definition and description considering it's not the resource or service being metered, but the associated SKUs (referring to 'SKU associated with a metered *resource* or *service*' vs 'SKU associated with a *resource* or *service*')
  - **Value ranges** – on hold.
    - Discuss and confirm before proceeding with the PR (table overview available in [#547: Corrections and value ranges - Metrics and ValueRanges sheet](https://docs.google.com/spreadsheets/d/1SvaD-PSzxwRKExKYbQXbtEmVPtbhN4HnycOatvLfGPI/edit?gid=1728767078#gid=1728767078)).
    - Suggestion: Switch to a bullet list before addressing value ranges.
  - A version with a bullet list and value ranges is available in a separate PR (not sure if we’ll manage to approve it by the 1.1 deadline).

**Comment by jpradocueva** [comment]
### Action Items from Members' call on September 26:
- [ ] [TF1- [#549](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/549) ] Members to review the changes and provide feedback, especially on value ranges and nullability.

File: specification/columns/consumedquantity.md [hunk 1]
```diff
@@ -1,14 +1,15 @@
 # Consumed Quantity
 
-The Consumed Quantity represents the volume of a given SKU associated with a metered [*resource*](#glossary:resource) or [*service*](#glossary:service) used, based on the [Consumed Unit](#consumedunit). Consumed Quantity is often derived at a finer granularity or over a different time interval when compared to the [Pricing Quantity](#pricingquantity) (complementary to [Pricing Unit](#pricingunit)) and focuses on metered *resource* and *service* consumption, not pricing and cost.
+The Consumed Quantity represents the volume of a metered SKU associated with a [*resource*](#glossary:resource) or [*service*](#glossary:service) used, based on the [Consumed Unit](#consumedunit). Consumed Quantity is often derived at a finer granularity or over a different time interval when compared to the [Pricing Quantity](#pricingquantity) (complementary to [Pricing Unit](#pricingunit)) and focuses on *resource* and *service* consumption, not pricing and cost.
```
Comment (cnharris10) [inline specification/columns/consumedquantity.md:3]: It seems off to me that we don't prominently include the word, `usage`, in a column dedicated to it.

```suggestion
The Consumed Quantity represents the volume usage of a metered SKU associated with a [*resource*](#glossary:resource) or [*service*](#glossary:service) used, based on the [Consumed Unit](#consumedunit). Consumed Quantity is often derived at a finer granularity or over a different time interval when compared to the [Pricing Quantity](#pricingquantity) (complementary to [Pricing Unit](#pricingunit)) and focuses on *resource* and *service* consumption, not pricing and cost.
```

File: specification/columns/consumedquantity.md [hunk 2]
```diff
@@ -20,7 +21,7 @@ Consumed Quantity
 
 ## Description
 
-The volume of a given SKU associated with a metered *resource* or *service* used, based on the Consumed Unit.
+The volume of a metered SKU associated with a *resource* or *service* used, based on the Consumed Unit.
```
Comment (cnharris10) [inline specification/columns/consumedquantity.md:24]:
```suggestion
The volume usage of a metered SKU associated with a *resource* or *service* used, based on the Consumed Unit.
```

File: specification/columns/consumedunit.md [hunk 3]
```diff
@@ -1,14 +1,16 @@
 # Consumed Unit
 
-The Consumed Unit represents a provider-specified measurement unit indicating how a provider measures usage of a given SKU associated with a metered [*resource*](#glossary:resource) or [*service*](#glossary:service). Consumed Unit complements the [Consumed Quantity](#consumedquantity) metric. It is often listed at a finer granularity or over a different time interval when compared to [Pricing Unit](#pricingunit) (complementary to [Pricing Quantity](#pricingquantity)), and focuses on metered *resource* and *service* consumption, not pricing and cost.
+The Consumed Unit represents a provider-specified measurement unit indicating how a provider measures usage of a metered SKU associated with a [*resource*](#glossary:resource) or [*service*](#glossary:service). Consumed Unit complements the [Consumed Quantity](#consumedquantity) metric. It is often listed at a finer granularity or over a different time interval when compared to [Pricing Unit](#pricingunit) (complementary to [Pricing Quantity](#pricingquantity)), and focuses on *resource* and *service* consumption, not pricing and cost.
```
Comment (cnharris10) [inline specification/columns/consumedunit.md:3]:
```suggestion
The Consumed Unit represents a provider-specified measurement unit indicating how a provider measures volume usage of a metered SKU associated with a [*resource*](#glossary:resource) or [*service*](#glossary:service). Consumed Unit complements the [Consumed Quantity](#consumedquantity) metric. It is often listed at a finer granularity or over a different time interval when compared to [Pricing Unit](#pricingunit) (complementary to [Pricing Quantity](#pricingquantity)), and focuses on *resource* and *service* consumption, not pricing and cost.
```

File: specification/columns/consumedunit.md [hunk 4]
```diff
@@ -20,7 +22,7 @@ Consumed Unit
 
 ## Description
 
-Provider-specified measurement unit indicating how a provider measures usage of a given SKU associated with a metered *resource* or *service*.
+Provider-specified measurement unit indicating how a provider measures usage of a metered SKU associated with a *resource* or *service*.
```
Comment (cnharris10) [inline specification/columns/consumedunit.md:25]:
```suggestion
Provider-specified measurement unit indicating how a provider measures volume usage of a metered SKU associated with a *resource* or *service*.
```

File: specification/columns/consumedquantity.md [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 
 The ConsumedQuantity column adheres to the following requirements:
 
 * ConsumedQuantity MUST be present in a FOCUS dataset when the provider supports the measurement of usage.
 * ConsumedQuantity MUST be of type Decimal and MUST conform to [Numeric Format](#numericformat) requirements.
-* ConsumedQuantity MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" and [ChargeClass](#chargeclass) is not "Correction".
-* ConsumedQuantity MUST be null if [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Unused" or for other *ChargeCategory* values.
-* ConsumedQuantity MAY be negative or null in cases where *ChargeClass* is "Correction".
+* If [ChargeCategory](#chargecategory) is "Usage" and [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Used", the following applies:
```
Comment (cnharris10) [inline specification/columns/consumedquantity.md:9]: How about this more simplified version?

```
* ConsumedQuantity MUST be null if [ChargeCategory](#chargecategory) is "Usage" and [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Unused".
* ConsumedQuantity MAY be negative or null in cases where [ChargeClass](#chargeclass) is "Correction".
* ConsumedQuantity MUST NOT be null in all other cases.
```

File: specification/columns/consumedunit.md [hunk 6]
```diff
@@ -1,14 +1,16 @@
...
 
 The ConsumedUnit column adheres to the following requirements:
 
 * ConsumedUnit MUST be present in the billing data when the provider supports the measurement of usage.
-* ConsumedUnit MUST be of type String, and the units of measure used in ConsumedUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.
-* ConsumedUnit MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" and [ChargeClass](#chargeclass) is not "Correction".
-* When *ChargeCategory* is not "Usage", or *ChargeCategory* is "Usage" and [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Unused", ConsumedUnit MUST be null when ChargeClass is not "Correction".
-* ConsumedUnit MAY be null if *ChargeCategory* is "Usage" and ChargeClass is "Correction".
+* ConsumedUnit MUST be of type String.
```
Comment (cnharris10) [inline specification/columns/consumedunit.md:8]: How about this more simplified version?
```
* ConsumedUnit MUST be of type String, and the units of measure used in ConsumedUnit SHOULD adhere to the values and format requirements specified in the UnitFormat attribute.
* ConsumedQuantity MUST be null if [ChargeCategory](#chargecategory) is not "Usage" or [ChargeCategory](#chargecategory) is "Usage" and [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Unused".
* ConsumedQuantity MAY be null when [ChargeClass](#chargeclass) is "Correction".
* ConsumedQuantity MUST NOT be null in all other cases.
```

**Comment by jpradocueva** [comment]
### Action Items from TF-1 call on October 1st:
- [ ] [TF1-[#549](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/549)] Irena, @ijurica : Implement the agreed minimum changes and prepare for future refinements.

File: specification/columns/consumedquantity.md (hunk 5 above)
Comment (ijurica) [reply specification/columns/consumedquantity.md:9]: There are some bugs in that suggestion, but I’ll follow the suggested principle and provide a new version.

File: specification/columns/consumedunit.md (hunk 6 above)
Comment (ijurica) [reply specification/columns/consumedunit.md]: I've provided new version following the suggested principle.

File: specification/columns/consumedquantity.md (hunk 5 above)
Comment (ijurica) [reply specification/columns/consumedquantity.md]: Chris' suggestion:
* ConsumedQuantity MUST be null unless [*ChargeCategory*](#chargecategory) is "Usage" and [CommitmentDiscountStatus](#commitmentdiscountstatus) is not "Unused".
* ConsumedQuantity MAY be negative or null in cases where [ChargeClass](#chargeclass) is "Correction".

**Comment by jpradocueva** [comment]
@cnharris10, @ijurica can you please review the comments and finalize them on time for tomorrow's Members Meeting, October 3rd? Please let me know if I can be of any assistance. Thanks!

**Review by udam-f2 — APPROVED** [review]

**Review by ljadvey — APPROVED** [review]

**Comment by jpradocueva** [comment]
Approved during Members' call on Oct 3rd.

**Pull request merged**